src/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs

```diff
@@ -232,6 +241,16 @@ public static object CreateDictionaryValue(ref ReadStack state)
         {
             JsonPropertyInfo jsonPropertyInfo = state.Current.JsonPropertyInfo;
 
+            if (jsonPropertyInfo == null)
```

Comment:
`jsonPropertyInfo` being `null` here means that the current type we expect to be parsed is not `ClassType.Enumerable` or `ClassType.Dictionary` (or `ClassType.Value`, but we don't expect this here because it's handled elsewhere). This means mismatched Json because we just saw a `StartArray` token.

If `jsonPropertyInfo.ClassType` is not `ClassType.Enumerable`, it means we have mismatched JSON: saw `StartArray` token, but we are processing a dictionary.